### PR TITLE
scrub bogus results when missing a clone

### PR DESCRIPTION
**Pull request by dzafman**
Fixes: 12738
Fixes: 12740

### CODE REVIEW

**Comment by dzafman** [comment]
Tests passed.

**Comment by dzafman** [comment]
I noticed an issue with this I want to fix and test again.

**Comment by tchaikov** [comment]
@dzafman it turns out to be longer than i expected to review this pull request. i will finish it tomorrow. this method has more than 1k LOC...

**Comment by tchaikov** [comment]
thanks @dzafman , your example helps a lot. it's like matching several concatenated zippers with missing teeth. .

**Comment by dachary** [comment]
@dzafman the bot false negative is hopefully the last occurrence of http://tracker.ceph.com/issues/13176 :-)

**Comment by dzafman** [comment]
@tchaikov @athanatos All updates now pushed here.

**Comment by dzafman** [comment]
@tchaikov Testing passed.  This is ready for final review.  If @athanatos isn't available maybe @liewegas can do it so there are 2 reviewers.

**Comment by dzafman** [comment]
@dachary Jenkins looks like all the tests passed, but a rocksdb error caused the failure.

**Comment by dachary** [comment]
@dzafman this is a known false negative, you can rebase and repush to try again  ( http://tracker.ceph.com/issues/13554 ).

File: src/osd/PG.cc [hunk 1]
```diff
@@ -3484,6 +3484,10 @@ void PG::_scan_snaps(ScrubMap &smap)
     if (hoid.snap < CEPH_MAXSNAP) {
       // fake nlinks for old primaries
       bufferlist bl;
+      if (o.attrs.find(OI_ATTR) == o.attrs.end()) {
+        o.nlinks = 1;
+	continue;
```
Comment (dzafman) [inline src/osd/PG.cc]: Now that I look at this I don't think a continue is correct.

File: src/osd/PG.cc [hunk 2]
```diff
@@ -3484,6 +3484,10 @@ void PG::_scan_snaps(ScrubMap &smap)
...
+      }
       bl.push_back(o.attrs[OI_ATTR]);
       object_info_t oi(bl);
```
Comment (dzafman) [inline src/osd/PG.cc]: We need to catch a decode failure here.

**Comment by dzafman** [comment]
I added DNM because I have commits to squash after review.

@athanatos @tchaikov See my additional 8 commits starting at 08ec091587ec9cb9a6f13074acd37db70e39ba1c

**Comment by dzafman** [comment]
I have to remove some asserts that snap != 0 and also an assumption that 0 can be used to mean all clones.

**Comment by dzafman** [comment]
@athanatos @tchaikov Testing passed.  When you both agree I'll squash and then merge this.

**Pull request merged**